### PR TITLE
ado-team-project option on generate-script

### PR DESCRIPTION
**Pull request by DevStarOps**
Adds ado-team-project option on generate-script #124 and updates service connection to be case insensitive as it was not finding the service connection in my org #242 

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked

### CODE REVIEW

File: src/Octoshift/AdoApi.cs [hunk 1]
```diff
@@ -85,7 +85,7 @@ public virtual async Task<string> GetGithubAppId(string org, string githubOrg, I
                 var url = $"https://dev.azure.com/{org}/{adoTeamProject}/_apis/serviceendpoint/endpoints?api-version=6.0-preview.4";
                 var response = await _client.GetWithPagingAsync(url);
 
-                var endpoint = response.FirstOrDefault(x => (string)x["type"] == "GitHub" && ((string)x["name"]).ToLower() == githubOrg.ToLower());
+                var endpoint = response.FirstOrDefault(x => ((string)x["type"]).Equals("GitHub", StringComparison.OrdinalIgnoreCase) && ((string)x["name"]).Equals(githubOrg, StringComparison.OrdinalIgnoreCase));
```
Comment (dylan-smith) [inline src/Octoshift/AdoApi.cs:88]: Is this related to the ado-team-project change, or is this just a bit of unrelated cleanup?

Comment (DevStarOps) [reply src/Octoshift/AdoApi.cs:88]: I created #242 because of a warning that is thrown at the end of the process where type returned is all lower case for me so was not matching GitHub

File: src/ado2gh/Commands/GenerateScriptCommand.cs [hunk 2]
```diff
@@ -161,16 +167,18 @@ public async Task Invoke(string githubOrg, string adoOrg, FileInfo output, bool
...
+                            await GetTeamProjectRepos(ado, repos, org, teamProject);
+                        }
+                    }
+                    else
+                    {
+                        if (teamProjects.Any(o => o.Equals(adoTeamProject, StringComparison.OrdinalIgnoreCase)))
                         {
-                            _log.LogInformation($"    Repo: {repo}");
+                            await GetTeamProjectRepos(ado, repos, org, adoTeamProject);
                         }
```
Comment (dylan-smith) [inline src/ado2gh/Commands/GenerateScriptCommand.cs:182]:
```suggestion
                            await GetTeamProjectRepos(ado, repos, org, adoTeamProject);
                        }
                        else
                        {
                            _log.LogWarning($"ADO Team Project provided cannot be found [{adoTeamProject}]");
                        }
```

File: src/ado2gh/Commands/GenerateScriptCommand.cs [hunk 3]
```diff
@@ -179,6 +187,18 @@ public async Task Invoke(string githubOrg, string adoOrg, FileInfo output, bool
             return repos;
         }
 
+        private async Task GetTeamProjectRepos(AdoApi ado, Dictionary<string, IDictionary<string, IEnumerable<string>>> repos, string org, string teamProject)
```
Comment (dylan-smith) [inline src/ado2gh/Commands/GenerateScriptCommand.cs:190]: I'd suggest returning an IEnumerable<string> and adding it to the repos Dict in the caller instead of passing in repos here.

File: src/gei/Commands/GenerateScriptCommand.cs [hunk 4]
```diff
@@ -87,7 +93,7 @@ public async Task Invoke(string githubSourceOrg, string adoSourceOrg, string git
             }
 
             var script = string.IsNullOrWhiteSpace(githubSourceOrg) ?
-                await InvokeAdo(adoSourceOrg, githubTargetOrg, ssh) :
+                await InvokeAdo(adoSourceOrg, adoTeamProject, githubTargetOrg, ssh) :
```
Comment (dylan-smith) [inline src/gei/Commands/GenerateScriptCommand.cs:96]: We should validate that the user hasn't passed in an Ado Team Project without an Ado Org, and throw an OctoshiftCliException if they did.

**Comment by dylan-smith** [comment]
I'm learning our CI doesn't work too well on PR's from forks. I'll file an issue to fix that later.

**Review by dylan-smith — CHANGES_REQUESTED** [review]

File: src/gei/Commands/GenerateScriptCommand.cs (hunk 4 above)
Comment (DevStarOps) [reply src/gei/Commands/GenerateScriptCommand.cs:96]: The check above says that if noth github org and ado org are empty it throws.
If both are supplied it goes the GitHub repos root, so if team project is supplied without org, then it's either missing GitHub org too and will throw or will exist

File: src/ado2gh/Commands/GenerateScriptCommand.cs (hunk 2 above)
Comment (DevStarOps) [reply src/ado2gh/Commands/GenerateScriptCommand.cs]: I thought about logging a warning but if you have 3 orgs that it's looping over and the team project exists in 1 of them then you'd get 2 warnings thinking something is wrong but it worked as expected

Comment (DevStarOps) [reply src/ado2gh/Commands/GenerateScriptCommand.cs]: What I'll do is capture if a team project was found and outside the org loop log the warning, same same, just a little more tracking code :)

File: src/OctoshiftCLI.Tests/AdoApiTests.cs [hunk 5]
```diff
@@ -238,6 +239,7 @@ public async Task GetGithubAppId_Should_Return_Null_When_No_Team_Projects_Have_E
             var githubOrg = "foo-gh-org";
             var teamProject1 = "foo-tp1";
             var teamProject2 = "foo-tp2";
+            var teamProjectArg = string.Empty;
```
Comment (dylan-smith) [inline src/OctoshiftCLI.Tests/AdoApiTests.cs:242]: depending on what we do with the previous comment, we may want a test for the scenario where teamProjectArg is not empty.

File: src/ado2gh/Commands/GenerateScriptCommand.cs [hunk 6]
```diff
@@ -171,36 +177,61 @@ public async Task Invoke(string githubOrg, string adoOrg, FileInfo output, bool
...
+                    _log.LogWarning($"ADO Team Project provided cannot be found [{adoTeamProject}]");
+                }
             }
 
             return repos;
         }
 
+        private async Task<IEnumerable<string>> GetTeamProjectRepos(AdoApi ado, string org, string teamProject)
+        {
+            _log.LogInformation($"  Team Project: {teamProject}");
```
Comment (dylan-smith) [inline src/ado2gh/Commands/GenerateScriptCommand.cs:225]: I think this is going to result in redundant log output with line 207

File: src/gei/Commands/GenerateScriptCommand.cs [hunk 7]
```diff
@@ -118,6 +124,7 @@ public GenerateScriptCommand(OctoLogger log, ISourceGithubApiFactory sourceGithu
             if (!string.IsNullOrWhiteSpace(adoSourceOrg))
             {
                 _log.LogInformation($"ADO SOURCE ORG: {adoSourceOrg}");
+                _log.LogInformation($"ADO TEAM PROJECT: {adoTeamProject}");
             }
```
Comment (dylan-smith) [inline src/gei/Commands/GenerateScriptCommand.cs:128]:
```suggestion
            }
            
            if (!string.IsNullOrWhiteSpace(adoTeamProject))
                _log.LogInformation($"ADO TEAM PROJECT: {adoTeamProject}");
            }
```

File: src/gei/Commands/GenerateScriptCommand.cs [hunk 8]
```diff
@@ -435,6 +448,18 @@ exit 1
             return content.ToString();
         }
 
+        private async Task<IEnumerable<string>> GetTeamProjectRepos(AdoApi adoApi, string adoOrg, string teamProject)
+        {
+            _log.LogInformation($"Team Project: {teamProject}");
```
Comment (dylan-smith) [inline src/gei/Commands/GenerateScriptCommand.cs:453]: duplicate log lines with line 232

**Review by dylan-smith — CHANGES_REQUESTED** [review]
The main change I think we need to make is how it handles finding the GH Pipelines App ID when scoped to a team project.  Once we decide that I can make all the other changes and merge it in.

File: src/ado2gh/Commands/GenerateScriptCommand.cs (hunk 6 above)
Comment (DevStarOps) [reply src/ado2gh/Commands/GenerateScriptCommand.cs:223]: updated, thanks, was the same on line 197 🙂

File: src/gei/Commands/GenerateScriptCommand.cs (hunk 8 above)
Comment (DevStarOps) [reply src/gei/Commands/GenerateScriptCommand.cs:456]: removed from earlier line so that both branches are covered by the method logging

**Comment by DevStarOps** [comment]
sorted all the issues, I think a lot of the duplicate logging I messed up during the merge as I added logging where I thought I was missing it forgetting I moved it to the method

For AppId I removed the restriction to only get the service connection for the team project supplied as that's meant to restrict repos and feels miss used and creates a harder migration process

**Review by dylan-smith — APPROVED** [review]
